src/Stratis.Bitcoin.Features.Consensus/ProvenBlockHeaders/ProvenBlockHeaderStore.cs

```diff
@@ -0,0 +1,374 @@
+
+            this.CurrentTipHashHeight = await this.CurrentTipHashHeightAsync().ConfigureAwait(false);
+
+            this.logger.LogDebug("Initialized ProvenBlockHader block tip at '{0}'.", this.CurrentTipHashHeight);
+
+            this.asyncLoop = this.asyncLoopFactory.Run("ProvenBlockHeaders job", token =>
+            {
+                this.logger.LogTrace("()");
+
+                // Save pending items.
+                this.SaveAsync().ConfigureAwait(false);
+
+                // Check and make sure the store cache limit isn't breached.
+                this.ManangeCacheSize();
+
+                this.logger.LogTrace("(-)[IN_ASYNC_LOOP]");
+
+                return Task.CompletedTask;
+            },
+            this.nodeLifetime.ApplicationStopping,
+            repeatEvery: TimeSpan.FromMinutes(1),
+            startAfter: TimeSpan.FromMinutes(1));
+        }
+
+        /// <inheritdoc />
+        public async Task<ProvenBlockHeader> GetAsync(int blockHeight)
+        {
+            ProvenBlockHeader header = null;
+
+            using (new StopwatchDisposable(o => this.performanceCounter.AddQueryTime(o)))
+            {
+                // Check pending cache first.
+                header = this.GetHeaderFromPendingCache(blockHeight);
+
+                if (header == null)
+                {
+                    // Check store cache next.
+                    header = this.GetHeaderFromStoreCache(blockHeight);
+
+                    if (header == null)
+                    {
+                        // Check the repository (DBreeze).
+                        header = await this.provenBlockHeaderRepository.GetAsync(blockHeight).ConfigureAwait(false);
+
+                        // Add the item to the store cache.
+                        this.Cache.AddOrUpdate(blockHeight, header);
+                    }
+                }
+            }
+
+            if (header != null)
+                this.logger.LogTrace("(-):*.{0}='{1}'", nameof(header), header);
+            else
+                this.logger.LogTrace("(-):null");
+
+            return header;
+        }
+
+        /// <inheritdoc />
+        public async Task<List<ProvenBlockHeader>> GetAsync(int fromBlockHeight, int toBlockHeight)
+        {
+            var cachedHeaders = new List<ProvenBlockHeader>();            
+
+            using (new StopwatchDisposable(o => this.performanceCounter.AddQueryTime(o)))
+            {
+                int heightKey = fromBlockHeight;
+
+                // Check if the items exist in the store cache before checking the repository.
+                ProvenBlockHeader headerCache = null;
+
+                var blockHeightList = new List<int>();
+
+                do
+                {
+                    headerCache = this.GetHeaderFromStoreCache(heightKey);
+
+                    if (headerCache != null)
+                        cachedHeaders.Add(headerCache);
```

Comment:
Perhaps a more correct name for the lists are headersInCache and headersNotInCache

Reply:
ok